### PR TITLE
New rule: prefer_foreach

### PR DESCRIPTION
**Pull request by JPaulsen**
Fixes https://github.com/dart-lang/linter/issues/634

### CODE REVIEW

**Comment by JPaulsen** [comment]
24 lints in the sdk, no false positives (Except the implementation of the forEach method itself), some examples:

```dart
for (var e in args) {
  list.add(e);
}
// Could be:
args.forEach(list.add);
```

```dart
for (var arg in args) {
  safeForTypedInterop(arg);
}
// Could be:
args.forEach(safeForTypedInterop);
```

```dart
for (E each in objects) {
  add(each);
}
// Could be:
objects.forEach(add);
```

File: lib/src/rules/prefer_foreach_call.dart [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+map.keys.toList().forEach(map.remove);
+```
+
+''';
+
+class PreferForeachCall extends LintRule {
+  _Visitor _visitor;
+  PreferForeachCall()
+      : super(
+            name: 'prefer_foreach_call',
```
Comment (bwilkerson) [inline lib/src/rules/prefer_foreach_call.dart]: I would drop the `_call`.

**Comment by JPaulsen** [comment]
@bwilkerson yes, you are totally right, I thought code like that would case errors so even I thought that the rule could have a little improvement in performance because of computing the Iterable just once.

So what do we do? What you said is totally right, I do not even know if we could computate if the change is safe so I think the options are: 

1.- Allow the rule as an experimental one.
2.- Discard it.

**Comment by alexeieleusis** [comment]
I like the rule, marking it as experimental would not fail analysis in CI and we can see how the community likes it, probably add a comment to the doc so besides being it marked as experimental make very explicit in docs that code relying on side effects should not perform this change.

File: lib/src/rules/prefer_foreach_call.dart (hunk 1 above)
Comment (JPaulsen) [inline lib/src/rules/prefer_foreach_call.dart]: Fixed

**Comment by JPaulsen** [comment]
@alexeieleusis done

**Comment by bwilkerson** [comment]
> So what do we do?

I suspect you already know my answer :-) I have a very low tolerance for lints with false positives, so I would lean toward dropping the rule. I won't stop you if you want to commit is as experimental, though.

**Comment by JPaulsen** [comment]
Thanks @bwilkerson

File: lib/src/rules/prefer_foreach.dart [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+for (final key in map.keys.toList()) {
+  map.remove(key);
+}
+```
+
+**GOOD:**
+```
+map.keys.toList().forEach(map.remove);
+```
+
```
Comment (alexeieleusis) [inline lib/src/rules/prefer_foreach.dart]: Add a comment here saying that the rule is experimental and that code that the suggested change alters logic if users rely on side effects over the iterable, provide an example.

Comment (JPaulsen) [inline lib/src/rules/prefer_foreach.dart]: Fixed

**Comment by JPaulsen** [comment]
@bwilkerson I was trying to figure out an example to show in the documentation but it looks like they iterable in the for each statement is being calculated just once so in that case is exactly the same behavior, I share this code to you https://dartpad.dartlang.org/c1c06ea89b842d6c822935f46bfe10a8

**Comment by bwilkerson** [comment]
By 'target' I meant the target of the method being invoked in the loop, not the iterator computed for the loop. Here's a comparison of the results of two methods, one named `process` that I think would produce a lint, and one named `processTransformed` that I think is what this lint would recommend transforming the code into:
```
main() {
  C c = new C();
  c.init();
  c.process(['a', 'b', 'c']);
  print(c.lists);

  c = new C();
  c.init();
  c.processTransformed(['a', 'b', 'c']);
  print(c.lists);
}

class C {
  var index = 0;
  var lists;

  next() => lists[index++];

  void init() {
    lists = [
    	[],
    	[],
    	[],
  	];
  }

  void process(list) {
    for (var element in list) {
      next().add(element);
    }
  }

  void processTransformed(list) {
    list.forEach(next().add);
  }
}
```
Note that the method `next` is invoked exactly once in the transformed code while it is invoked multiple times in the loop-form.

If the lint won't be produced for the loop in `perform` because the target can't be proven to not have any side-effects, then I'm happy with the rule as is. But if the lint *will* be produced, then there's a chance for false positives.

**Comment by JPaulsen** [comment]
@alexeieleusis comment fixed and added

**Comment by alexeieleusis** [comment]
We can avoid those false positives by linting only on local variables or fields, or said differently not linting method invocations and getters.

I will do that change later.

**Pull request merged**